### PR TITLE
Deprecate peapod, provide migration to fstree

### LINKED ISSUE #2924: Deprecate peapod, provide migration to fstree
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we keep code we no longer need. After #2814 peapods can be considered obsolete since fstree has better performance in all cases (and can be tweaked in many ways). We also:
 * have #2813 that is peapod-specific
 * have #1951 that peapods are susceptible to
 * have #2107 that is not peapod-specific, but likely can be handled much easier without yet another BoltDB on the disk
 * have #2497 where we can finally eliminate the concept of substorages, one shard --- one blobstor is much easier to handle (not within the scope of current issue, but in future, see #2480 also)

## Describe the solution you'd like
The proposal is to discontinue them in the same way we did for bbcz:
 * declare them as deprecated and to be removed
 * create a migration tool that moves data and fixes configuration file
 * roll it all out, keep peapods supported for one release, migrate
 * drop support for peapods in subsequent releases

## Describe alternatives you've considered
Keeping peapods, but why?

### PR DESCRIPTION
**Pull request by End-rey**
Closes #2924.

Is there anything else I should do in this pr?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -59,6 +61,23 @@ introduced in version 0.22.3 and support for binary keys was removed from
...
+For any shard, the data from the configured Peapod is copied into
+a FSTree that must be already configured, if it is not, into
+a created FSTree dir named `blobstore` in the directory where the tree is
+located. For example, `/neofs/data/peapod.db` -> `/neofs/data/blobstore`.
+Notice that existing Peapod are untouched. Configuration is also
+updated, for example, `/etc/neofs/config.yaml` -> `/etc/neofs/config_fstree.yaml`.
+WARN: carefully review the updated config before using it in the application!
+
+Now there is an FSTree in which files of any size will be more efficient and easier to store.
+Use only `fstree` sub-storage in `blobstor` config section.
```
Comment (roman-khimov) [inline CHANGELOG.md:77]: I'd say that currently we support both, but peapod can be removed in future versions. And we recommend using fstree.

File: CHANGELOG.md [hunk 2]
```diff
@@ -59,6 +61,23 @@ introduced in version 0.22.3 and support for binary keys was removed from
...
 
+To migrate data from Peapods to FSTree:
+```shell
+$ peapod-to-fstree -config </path/to/storage/node/config>
+```
+For any shard, the data from the configured Peapod is copied into
+a FSTree that must be already configured, if it is not, into
+a created FSTree dir named `blobstore` in the directory where the tree is
+located. For example, `/neofs/data/peapod.db` -> `/neofs/data/blobstore`.
+Notice that existing Peapod are untouched. Configuration is also
```
Comment (roman-khimov) [inline CHANGELOG.md:72]: peapod DB is not deleted during migration

File: pkg/local_object_storage/blobstor/peapod/peapod.go [hunk 3]
```diff
@@ -39,6 +39,8 @@ type batch struct {
 // Peapod provides storage for relatively small NeoFS binary object (peas).
 // Peapod is a single low-level key/value database optimized to work with big
 // number of stored units.
+//
+// Deprecated: Use fstree.FSTree instead.
```
[the remark this answers is not in the document]
Comment (End-rey) [reply pkg/local_object_storage/blobstor/peapod/peapod.go:43]: What about `peapod.New`? Maybe I should leave him too?

File: cmd/peapod-to-fstree/main.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+			return nil
+		}
+
+		ppdPath := ppd.Path()
+		if !filepath.IsAbs(ppdPath) {
+			log.Fatalf("Peapod path '%s' is not absolute, make it like this in the config file first\n", ppdPath)
+		}
+
+		if fstr == nil {
+			fstr = fstree.New(
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply cmd/peapod-to-fstree/main.go:78]: Right now the limits imposed on config (2 substorages min) are suggesting that we can't have a blobstor without fstree. And I don't like this magic in general, we can legitimately fail in this case, just say "please configure some fstree for this shard" and let the user figure it out. This won't be a problem in real life.

File: pkg/local_object_storage/blobstor/peapod/peapod.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/blobstor/peapod/peapod.go:43]: dont we (and this exact commit) need to print REALLY HUGE LETTERS IN LOGS PLEASE DO NOT USE PEAPOD ANYMORE?

File: CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,7 @@ attribute, which is used for container domain name in NNS contracts (#2954)
 - Docs files for cli commands to the `docs/cli-commands` folder (#2983)
 - `logger.encoding` config option (#2999)
 - Reloading morph endpoints with SIGHUP (#2998)
+- New `peapod-to-fstree` tool providing peapod-to-fstree data migration (#3013)
```
Comment (carpawell) [inline CHANGELOG.md:22]: this commit's message was ruined but some squash? should be fixed

File: cmd/peapod-to-fstree/main.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+					storagesCfg[i].Perm(),
+					storagesCfg[i].FlushInterval(),
+				)
+			default:
+				return fmt.Errorf("invalid storage type: %s", storagesCfg[i].Type())
+			}
+		}
+
+		if ppd == nil {
+			log.Println("Peapod is not configured for the current shard, going to next one...")
```
Comment (carpawell) [inline cmd/peapod-to-fstree/main.go:69]: for what shard?

File: cmd/peapod-to-fstree/main.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	flag.Parse()
+
+	if *nodeCfgPath == "" {
+		log.Fatal("missing storage node config flag")
+	}
+
+	appCfg := config.New(config.Prm{}, config.WithConfigFile(*nodeCfgPath))
+
+	err := engineconfig.IterateShards(appCfg, false, func(sc *shardconfig.Config) error {
+		log.Println("processing shard...")
```
Comment (carpawell) [inline cmd/peapod-to-fstree/main.go:36]: what shard?

**Review by carpawell — CHANGES_REQUESTED** [review]
I am almost sure we also want to migrate some metabase indexes _in this util_, otherwise i do not know who should do it.
See https://github.com/nspcc-dev/neofs-node/blob/ba1d63952ef555989171bb3e017fe054f3abceb2/pkg/local_object_storage/blobstor/get.go#L19-L32, https://github.com/nspcc-dev/neofs-node/blob/ba1d63952ef555989171bb3e017fe054f3abceb2/pkg/local_object_storage/blobstor/peapod/peapod.go#L422-L425 and https://github.com/nspcc-dev/neofs-node/blob/ba1d63952ef555989171bb3e017fe054f3abceb2/pkg/local_object_storage/blobstor/fstree/fstree.go#L308

**Comment by carpawell** [comment]
Linter feels bad.

File: pkg/local_object_storage/blobstor/peapod/peapod.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/local_object_storage/blobstor/peapod/peapod.go:43]: No. Not yet.

File: cmd/peapod-to-fstree/main.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/blobstor/peapod"
+	meta "github.com/nspcc-dev/neofs-node/pkg/local_object_storage/metabase"
+	"go.etcd.io/bbolt"
+	"gopkg.in/yaml.v3"
+)
+
+type epochState struct{}
+
+func (s epochState) CurrentEpoch() uint64 {
+	return 0
```
Comment (roman-khimov) [inline cmd/peapod-to-fstree/main.go:32]: Not sure what can be affected by this.

Comment (End-rey) [reply cmd/peapod-to-fstree/main.go:32]: How can I easily get the epoch state in another way? The metabase requires it to be set.

Comment (roman-khimov) [reply cmd/peapod-to-fstree/main.go:32]: Only via network interaction. Maybe it's not a problem in our case (updating storage ID doesn't require epoch IIRC). Or maybe we can get some latest one from meta or local state.

Comment (End-rey) [reply cmd/peapod-to-fstree/main.go:32]: I can get LastResyncEpoch from metabase:
https://github.com/nspcc-dev/neofs-node/blob/45802c65709a8be508ccd8b949efaf85933ddd6f/pkg/local_object_storage/metabase/last_resync_epoch.go#L12-L14

Comment (roman-khimov) [reply cmd/peapod-to-fstree/main.go:32]: It's somewhat different in meaning and it's a new one, 0.43.0 DB doesn't have it.

Comment (carpawell) [reply cmd/peapod-to-fstree/main.go:32]: why do you need a precise epoch? just pick an epoch that does not claim anything about object expiration and allows GET and PUT freely, that is `0`. @roman-khimov, what bothers you?

> How can I easily get the epoch state in another way? The metabase requires it to be set.

i did this and i did not know there would be so many utils that just open metabases and look at them. normal metabase always has to know the current epoch (via internal state or via explicit parameter to every operation)

**Review by roman-khimov — APPROVED** [review]

File: cmd/peapod-to-fstree/main.go [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+	nodeCfgPath := flag.String("config", "", "Path to storage node's YAML configuration file")
+
+	flag.Parse()
+
+	if *nodeCfgPath == "" {
+		log.Fatal("missing storage node config flag")
+	}
+
+	srcPath := *nodeCfgPath
+	ss := strings.Split(srcPath, ".")
```
Comment (carpawell) [inline cmd/peapod-to-fstree/main.go:45]: i would recomment smth like
```go
extension := filepath.Ext(srcPath)
name := srcPath[0:len(srcPath)-len(extension)]
```

dots are not prohibited in file names

**Review by carpawell — APPROVED** [review]

File: cmd/peapod-to-fstree/main.go (hunk 9 above)
Comment (roman-khimov) [reply cmd/peapod-to-fstree/main.go:45]: In general, you can just do `+ ".migrated"` and that's it. `srcPath` can be anything, with many dots, without dots, whatever.

**Pull request merged**